src/core/src/package_managers/ZypperPackageManager.py

```diff
@@ -44,71 +44,96 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
-        return out
+
+        for i in range(0, self.package_manager_max_retries):
+            zypp_lock_timeout_backup = None
+            try:
+                zypp_lock_timeout_backup = self.env_layer.get_env_var('ZYPP_LOCK_TIMEOUT')
+                self.composite_logger.log_debug("Original value of ZYPP_LOCK_TIMEOUT env var: {0}".format(str(zypp_lock_timeout_backup)))
+                self.env_layer.set_env_var('ZYPP_LOCK_TIMEOUT', 5)
+
+                code, out = self.env_layer.run_command_output(command, False, False)
+                if code not in [self.zypper_exitcode_ok, self.zypper_exitcode_zypper_updated]:  # more known return codes should be added as appropriate
+                    self.composite_logger.log('[ERROR] Package manager was invoked using: ' + command)
+                    self.composite_logger.log_warning(" - Return code from package manager: " + str(code))
+                    self.composite_logger.log_warning(" - Output from package manager: \n|\t" + "\n|\t".join(out.splitlines()))
+
+                    process_tree = self.get_process_tree_from_pid_in_output(out)
+                    if process_tree is not None:
+                        self.composite_logger.log_warning(" - Process tree for the pid in output: \n{}".format(str(process_tree)))
+
+                    self.telemetry_writer.write_execution_error(command, code, out)
+                    error_msg = 'Unexpected return code (' + str(code) + ') from package manager on command: ' + command
+
+                    if code in self.error_codes_to_retry:
+                        raise Exception(error_msg)
+                    else:
+                        self.status_handler.add_error_to_status(error_msg, Constants.PatchOperationErrorCodes.PACKAGE_MANAGER_FAILURE)
+                        raise Exception(error_msg, "[{0}]".format(Constants.ERROR_ADDED_TO_STATUS))
+                else:  # verbose diagnostic log
+                    self.composite_logger.log_debug("\n\n==[SUCCESS]===============================================================")
+                    self.composite_logger.log_debug(" - Return code from package manager: " + str(code))
+                    self.composite_logger.log_debug(" - Output from package manager: \n|\t" + "\n|\t".join(out.splitlines()))
+                    self.composite_logger.log_debug("==========================================================================\n\n")
+
+                if code == self.zypper_exitcode_zypper_updated:
+                    self.composite_logger.log_debug(" - Package manager update detected. Patch installation run will be repeated.")
+                    self.set_package_manager_setting(Constants.PACKAGE_MGR_SETTING_REPEAT_PATCH_OPERATION, True)
+                return out
+            except Exception as error:
+                error_msg = repr(error)
+                if Constants.ERROR_ADDED_TO_STATUS in error_msg:
+                    raise
```

Comment:
Looks like we are depending on Constants.ERROR_ADDED_TO_STATUS to determine if retries are required, which does not look right.

 Constants.ERROR_ADDED_TO_STATUS is used to decide whether or not the error is to be added to error objects and should not be used for other cases. Better to have the usages independent, so that in future if the implementation for error objects change, it won't affect these retries.

Reply:
Understood. I wanted some way to share the output code between the try/except blocks and opted for that. I have now just moved the code outside of those blocks so it's accessible to both.